File: ijdc_parser.php
<?php

// ============================================================================
// IJDC Parser - Simple, Clean
// ============================================================================


// ============================================================================
// Parse IJDC File
// Read and store structure as-is
// ============================================================================

function parseIJDC($filePath) {
    if (!file_exists($filePath)) {
        return ['error' => 'File not found'];
    }

    $lines = file($filePath, FILE_IGNORE_NEW_LINES);
    $data = [];
    $inBlock = false;
    $blockKey = '';
    $blockContent = '';

    foreach ($lines as $index => $line) {
        $line = rtrim($line);

        // Skip version on first line
        if ($index === 0 && strpos($line, 'version:ijdc') === 0) {
            $data['_version'] = trim(substr($line, 12));
            continue;
        }

        // Skip empty lines
        if (empty($line)) {
            continue;
        }

        // Opening a block
        if (strpos($line, '<<^') !== false && !$inBlock) {
            $colonPos = strpos($line, ':');
            $blockKey = substr($line, 0, $colonPos);

            // Block closes on same line
            if (strpos($line, '^>>') !== false) {
                $start = strpos($line, '<<^') + 3;
                $end = strpos($line, '^>>');
                $blockContent = substr($line, $start, $end - $start);

                if (substr($blockKey, 0, 1) !== '-') {
                    if (isset($data[$blockKey]) && !is_array($data[$blockKey])) {
                        $data[$blockKey] = [$data[$blockKey]];
                    }
                    if (isset($data[$blockKey])) {
                        $data[$blockKey][] = $blockContent;
                    } else {
                        $data[$blockKey] = $blockContent;
                    }
                }
            } else {
                // Block continues
                $inBlock = true;
                $start = strpos($line, '<<^') + 3;
                $blockContent = substr($line, $start);
            }
            continue;
        }

        // Closing a block
        if ($inBlock && strpos($line, '^>>') !== false) {
            $end = strpos($line, '^>>');
            $blockContent .= "\n" . substr($line, 0, $end);
            $inBlock = false;

            if (substr($blockKey, 0, 1) !== '-') {
                if (isset($data[$blockKey]) && !is_array($data[$blockKey])) {
                    $data[$blockKey] = [$data[$blockKey]];
                }
                if (isset($data[$blockKey])) {
                    $data[$blockKey][] = trim($blockContent);
                } else {
                    $data[$blockKey] = trim($blockContent);
                }
            }
            continue;
        }

        // Inside a block
        if ($inBlock) {
            $blockContent .= "\n" . $line;
            continue;
        }

        // Scalar field
        if (strpos($line, ':') !== false) {
            $colonPos = strpos($line, ':');
            $key = substr($line, 0, $colonPos);
            $value = substr($line, $colonPos + 1);

            if (substr($key, 0, 1) !== '-') {
                if (isset($data[$key]) && !is_array($data[$key])) {
                    $data[$key] = [$data[$key]];
                }
                if (isset($data[$key])) {
                    $data[$key][] = $value;
                } else {
                    $data[$key] = $value;
                }
            }
        }
    }

    return $data;
}


// ============================================================================
// Process Content Field
// Convert (^( to <pre> and )^) to </pre>
// Convert code tags back to normal
// ============================================================================

function processContent($content) {
    $content = str_replace('(^(', '<pre>', $content);
    $content = str_replace(')^)', '</pre>', $content);
    $content = str_replace('|code|||~php~', '&lt;?php', $content);
    $content = str_replace('~php~|||code|', '?&gt;', $content);
    $content = str_replace('|code|||~!DOCTYPE html~', '&lt;!DOCTYPE html&gt;', $content);
    $content = str_replace('|code|||~html~', '&lt;html&gt;', $content);
    $content = str_replace('~/html~|||code|', '&lt;/html&gt;', $content);
    $content = str_replace('|code|||~head~', '&lt;head&gt;', $content);
    $content = str_replace('~/head~|||code|', '&lt;/head&gt;', $content);
    $content = str_replace('|code|||~body~', '&lt;body&gt;', $content);
    $content = str_replace('~/body~|||code|', '&lt;/body&gt;', $content);
    $content = str_replace('|code|||~', '&lt;', $content);
    $content = str_replace('~|||code|', '&gt;', $content);
    return $content;
}

$ijdcFile = __DIR__ . '/sample2/recipe_chat.ijdc';
$ijdcData = parseIJDC($ijdcFile);

if (isset($ijdcData['error'])) {
    echo 'Error: ' . htmlspecialchars($ijdcData['error']);
    return;
}

?>
<hr>
	<?php
		$before = " • ";       						// html before the link
		$after = " • ";             							// html after the link
		$show_submenu = 0; 
		//$thisparent	= 52;	// show the sub menu   1        or not  0
     // record id numbers the   AREA(id) 
		include ('./securecabinet/site_nav.bullet');
?>
<hr>
<?php foreach ($ijdcData as $key => $value): ?>

<?php if ($key === '_version' || $key === 'title' || $key === 'conversation_id' || $key === 'created_at' || $key === 'updated_at' || $key === 'conversation_mode' || $key === 'model') continue; ?>

<div>
    <strong><?php echo htmlspecialchars($key); ?>:</strong>

    <?php if (is_array($value)): ?>

        <?php foreach ($value as $item): ?>

            <div style="margin-left: 20px; margin-top: 10px;">
                <?php
                // Check if item is piped sub-fields (has | and key:value pattern)
                if (strpos($item, '|') !== false && strpos($item, ':') !== false && strpos($item, '|code|||') === false) {
                    // Parse sub-fields
                    $subFields = explode('|', $item);
                    foreach ($subFields as $subField) {
                        $subField = trim($subField);
                        if (strpos($subField, ':') !== false) {
                            $colonPos = strpos($subField, ':');
                            $subKey = trim(substr($subField, 0, $colonPos));
                            $subValue = trim(substr($subField, $colonPos + 1));
                            
                            echo '<div style="margin-bottom: 10px;">';
                            echo '<strong>' . htmlspecialchars($subKey) . ':</strong> ';
                            
                            // Check if subValue has code blocks
                            if (strpos($subValue, '(^(') !== false || strpos($subValue, '|code|||~') !== false) {
                                echo processContent($subValue);
                            } else {
                                echo htmlspecialchars($subValue);
                            }
                            echo '</div>';
                        }
                    }
                } else {
                    // Large content block - check for code or pre
                    if (strpos($item, '(^(') !== false || strpos($item, '|code|||~') !== false) {
                        echo processContent($item);
                    } else {
                        echo htmlspecialchars($item);
                    }
                }
                ?>
            </div>

        <?php endforeach; ?>

    <?php else: ?>

        <div style="margin-left: 20px;">
            <?php echo htmlspecialchars($value); ?>
        </div>

    <?php endif; ?>

</div>

<div style="margin-top: 15px;"></div>

<?php endforeach; ?>

File: ijdc_parser.py
from pathlib import Path
from typing import Dict, Any, List, Union
import re

# --------->>>>>>>  THIS PARSER WORKS JDC March 03 2026 16:33
def parse_ijdc(file_path: str | Path) -> Dict[str, Any]:
    file_path = Path(file_path)
    if not file_path.is_file():
        return {"error": "File not found"}

    data: Dict[str, Any] = {}
    in_block = False
    block_key = ""
    block_content_lines: List[str] = []

    with file_path.open("r", encoding="utf-8") as f:
        lines = f.readlines()

    for line in lines:
        line_raw = line.rstrip("\r\n")

        if in_block:
            if "^>>" in line_raw:
                end = line_raw.find("^>>")
                block_content_lines.append(line_raw[:end])
                full_content = "".join(block_content_lines)
                _append(data, block_key, full_content)
                in_block = False
                block_content_lines = []
            else:
                block_content_lines.append(line)
            continue

        colon_pos = line_raw.find(":")
        if colon_pos == -1:
            continue

        key = line_raw[:colon_pos].rstrip()
        value_start = line_raw[colon_pos + 1:]

        if value_start.lstrip().startswith("<<^"):
            block_key = key
            start = value_start.find("<<^") + 3
            if "^>>" in value_start:
                end = value_start.find("^>>", start)
                content = value_start[start:end]
                _append(data, key, content)
            else:
                in_block = True
                block_content_lines = [value_start[start:] + "\n"]
        else:
            _append(data, key, value_start)

    return data


def _append(data: Dict[str, Any], key: str, value: Any) -> None:
    if key in data:
        if not isinstance(data[key], list):
            data[key] = [data[key]]
        data[key].append(value)
    else:
        data[key] = value


def process_code_block(content: str) -> str:
    content = content.removeprefix("(^(").removesuffix(")^)")

    content = content.replace("&lt;", "<").replace("&gt;", ">").replace("&quot;", '"')

    replacements = [
        ('|code|||~php~', '<?php'),
        ('~php~|||code|', '?>'),
        ('|code|||~!DOCTYPE html~', '<!DOCTYPE html>'),
        ('|code|||~html~', '<html>'),
        ('~/html~|||code|', '</html>'),
        ('|code|||~head~', '<head>'),
        ('~/head~|||code|', '</head>'),
        ('|code|||~body~', '<body>'),
        ('~/body~|||code|', '</body>'),
        ('|code|||~', '<'),
        ('~|||code|', '>'),
    ]
    for old, new in replacements:
        content = content.replace(old, new)

    return content


def render_message_block(item: str) -> None:
    # Find the prefix up to content:
    # We split only the part before the content block
    content_start = item.find("content:")
    if content_start == -1:
        print(item.strip())
        return

    prefix = item[:content_start].strip()
    content_part = item[content_start + 8:].strip()  # after "content:"

    # Split prefix on | (role and timestamp)
    prefix_parts = [p.strip() for p in prefix.split("|") if p.strip()]

    role = ""
    timestamp = ""
    for part in prefix_parts:
        if ":" not in part:
            continue
        k, v = [x.strip() for x in part.split(":", 1)]
        if k == "role":
            role = v
        elif k == "timestamp":
            timestamp = v

    print(f"role: {role}")
    print(f"timestamp: {timestamp}")
    print("content: ", end="")

    # Now process the full content part
    if content_part.startswith("(^("):
        print()  # newline before code block
        print(process_code_block(content_part))
    else:
        print(content_part)


def render_item(item: str) -> None:
    stripped = item.strip()

    if "role:" in stripped and "timestamp:" in stripped and "content:" in stripped:
        render_message_block(item)
    else:
        if stripped.lstrip().startswith("(^("):
            print(process_code_block(item))
        else:
            print(item.rstrip("\r\n"))


if __name__ == "__main__":
    script_dir = Path(__file__).parent.resolve()
    ijd_file = script_dir / "recipe_chat.ijdc"

    parsed = parse_ijdc(ijd_file)

    if "error" in parsed:
        print("Error:", parsed["error"])
        print("Current dir:", script_dir)
    else:
        #print("=== Parsed IJDC ===")
        print()

        skip = {"version","dev_comment", "title", "conversation_id", "created_at", "updated_at", "conversation_mode", "model"}

        for key, val in parsed.items():
            if key in skip:
                continue

            print(f"{key}:")

            if isinstance(val, list):
                for item in val:
                    render_item(item)
                    print()
            else:
                render_item(val)

            #print("-" * 60)
            print()